cwl/src/main/scala/cwl/WorkflowStep.scala

```diff
@@ -281,36 +352,131 @@ case class WorkflowStep(
+        })
+          .sequence[ErrorOr, (String, ExpressionNode)]
+          .toEither
+          .map(_.toMap)
+      }
+
       //inputs base case consist of the nodes we already know about
       val baseCase = WorkflowStepInputFold(generatedNodes = knownNodes).asRight[NonEmptyList[String]]
 
       // WorkflowStepInputFold contains the mappings from step input to ExpressionNode as well as all created nodes
       val stepInputFoldCheck: Checked[WorkflowStepInputFold] = in.foldLeft(baseCase)(foldStepInput)
 
       /*
-        1) Fold over the workflow step inputs:
-          - Create an expression node for each input
-          - recursively generates unseen call nodes as we discover them going through step input sources
-          - accumulate all that in the WorkflowStepInputFold
-        2) Fold over the callable input definition using the expression node map from 1):
-          - determine the correct mapping for the input definition based on the expression node map
-          and the type of input definition
-          - accumulate those mappings, along with potentially newly created graph input nodes as well as call input ports
-          in an InputDefinitionFold
-        3) Use the InputDefinitionFold to build a new call node
+        * This (big) flatMap builds nodes from top to bottom in the diagram above.
+        * If necessary, the scatter node is built last as it wraps some of the other nodes.
        */
       for {
+        /* ************************************ */
+        /* ************ Merge Nodes *********** */
+        /* ************************************ */
+        // Build merge nodes and recursively generates other call nodes that we haven't seen so far
         stepInputFold <- stepInputFoldCheck
+        // Extract the merge nodes from the fold
+        mergeNodes = stepInputFold.mergeNodes
+
+        /* ************************************ */
+        /* ****** Scatter Variable Nodes ****** */
+        /* ************************************ */
+        scatterVariableNodes <- ScatterLogic.buildScatterVariableNodes(scatter, mergeNodes, unqualifiedStepId.localName.value)
+
+        /* ************************************ */
+        /* *************** OGINS ************** */
+        /* ************************************ */
+        ogins = buildOGINs(mergeNodes, scatterVariableNodes)
+
+        /* ************************************ */
+        /* ********* Expression Nodes ********* */
+        /* ************************************ */
+        // Aggregate the generated nodes so far. This map will be used to generate expression nodes, so the order of aggregation matters:
+        // scatter variables and ogins take precedence over merge nodes (see diagram)
+        expressionNodeMap = mergeNodes ++ scatterVariableNodes ++ ogins
+        // Build expression nodes for inputs that have a valueFrom field
+        stepInputExpressionNodes <- buildStepInputExpressionNodes(expressionNodeMap)
```

Comment:
The comment makes it clear what's going on here, but I wonder if giving the variable a name that contains 'valueFrom' might make what happens below easier to follow.